vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs

```diff
@@ -0,0 +1,110 @@
+        projectInfoManager: ProjectInfoManager
+    ) =
+
+    static member GetFormattingChanges(document: Document, sourceText: SourceText, checker: FSharpChecker, optionsOpt: FSharpProjectOptions option, position: int) =
+        // Logic for determining formatting changes:
+        // If first token on the current line is a closing brace,
+        // match the indent with the indent on the line that opened it
+
+        asyncMaybe {
+            
+            // Gate formatting on whether smart indentation is enabled
+            // (this is what C# does)
+            do! Option.guard (FSharpIndentationService.IsSmartIndentEnabled document.Project.Solution.Workspace)
+
+            let! options = optionsOpt
+            
+            let line = sourceText.Lines.[sourceText.Lines.IndexOf position]
+                
+            let defines = CompilerEnvironment.GetCompilationDefinesForEditing(document.FilePath, options.OtherOptions |> List.ofArray)
+
+            let tokens = Tokenizer.tokenizeLine(document.Id, sourceText, line.Start, document.FilePath, defines)
+
+            let! firstMeaningfulToken = 
+                tokens
+                |> List.tryFind (fun x ->
+                    x.Tag <> FSharpTokenTag.WHITESPACE &&
+                    x.Tag <> FSharpTokenTag.COMMENT &&
+                    x.Tag <> FSharpTokenTag.LINE_COMMENT)
+
+            let! (left, right) =
+                FSharpBraceMatchingService.GetBraceMatchingResult(checker, sourceText, document.FilePath, options, position, "FormattingService")
+
+            if right.StartColumn = firstMeaningfulToken.LeftColumn then
+                // Replace the indentation on this line with the indentation of the left bracket
+                let! leftSpan = RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, left)
+
+                let indentChars (line : TextLine) =
+                    line.ToString()
+                    |> Seq.takeWhile ((=) ' ')
+                    |> Seq.length
+                    
+                let startIndent = indentChars sourceText.Lines.[sourceText.Lines.IndexOf leftSpan.Start]
+                let currentIndent = indentChars line
+
+                return TextChange(TextSpan(line.Start, currentIndent), String.replicate startIndent " ")
+            else
+                return! None
+        }
+
+    member __.GetFormattingChangesAsync (document: Document, position: int, cancellationToken: CancellationToken) =
+        async {
+            let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+            let optionsOpt = projectInfoManager.TryGetOptionsForEditingDocumentOrProject document
+            let! textChange = FSharpEditorFormattingService.GetFormattingChanges(document, sourceText, checkerProvider.Checker, optionsOpt, position)
+                
+            return
+                match textChange with
+                | Some change ->
+                    List([change]) :> IList<_>
```

Comment:
We normally use the name ``ResizeArray`` in F# code, but it's no big deal (not block merge on this)